### ISSUE TITLE
Is metadata push watertight?

### ISSUE THREAD

**Issue by Sophietje** [TNO/knowledge-engine#134]
In GitLab by @barry.nouwt.tno.nl on Mar 5, 2021, 08:22

This means that there is no possibility that the information of a Knowledge Base is incomplete and will remain incomplete due to some race condition.

**Comment by Sophietje** [comment]
In GitLab by @barry.nouwt.tno.nl on Mar 26, 2021, 16:44

Apparently not, when running Thermostat app the following message appeared:

```
[pool-4-thread-2] WARN https%3A%2F%2Fwww.tno.nl%2Fthermostat-OtherKnowledgeBaseStoreImpl - Tried to update knowledge base https://www.tno.nl/heatingSource, but it is not in my store! Skipped it.
```

**Comment by Sophietje** [comment]
In GitLab by @barry.nouwt.tno.nl on Mar 29, 2021, 16:34

Solved the previous issue by fixing a bug in the OtherKnowledgeBaseStoreImpl. Now a new problem occurs where the ordering of messages send and received becomes important. The problems happens in some of the AskAnswerTests where an earlier metadata message is processed after a later metadata message and the later metadata gets overwritten by the earlier metadata which misses some of the Knowledge Interactions.

**Comment by Sophietje** [comment]
In GitLab by @barry.nouwt.tno.nl on Apr 19, 2021, 15:27

These AskAnswerTests seem to have been fixed as well, but maybe not that elegant. We have made the register method wait until it has received a confirmation from the receiving knowledge base. This prevents the first register to arrive after the second register.

**Comment by Sophietje** [comment]
In GitLab by @barry.nouwt.tno.nl on Apr 23, 2021, 13:38

Solved with a quick fix #151, but a more elaborate solution using #169 should happen at some point.
